job still running though received kill signal

### Nomad version
0.9.0-beta3

### Operating system and Environment details
OS: centos
OS version: 7.6.1810
### Issue
driver [exec]
the job jobA: one per node.
I clicked the stop button in the nomad-ui，some are [completed] immediately, but others are still [running] though the job status is dead. like this:

![nomad](https://user-images.githubusercontent.com/19262116/53847663-881dd900-3fec-11e9-98cc-a49bfce195b7.PNG)


here are the events of which is still running:

![nomad1](https://user-images.githubusercontent.com/19262116/53847786-f6629b80-3fec-11e9-9e8c-4e307b785c5e.PNG)

log in the nomad client:
2019-03-06T08:56:52.296+0800 [ERROR] client.driver_mgr.exec: error receiving stream from Stats executor RPC, closing stream:alloc_id=ce203f54-78cd-fa32-aff0-391c20e219ba driver=exec task_name=artifact error="rpc error: code = Unavailable desc = transport is closing"

when I kill  the process manually, it,s still running.

when I clicked the start button, the node which has the jobA still running will has a pending jobA. like this:
![nomad2](https://user-images.githubusercontent.com/19262116/53848803-aede0e80-3ff0-11e9-8c41-82ac222ea42e.PNG)

the /system/gc or client/allocation/id/gc api doesn.t work.


when I remove the node which has the still running jobA and gc. somehow it all works fine.

@preetapan  it,s hard to reproduce, and this could happen on different job. this time is the job: promtail. 
when I click the stop button on the nomad-ui:

Here is the log from nomad client:

2019-03-11T12:17:06.863+0800 [ERROR] client.driver_mgr.raw_exec: error receiving stream from Stats executor RPC, closing stream: alloc_id=6d3679d1-f40e-f2ce-080c-3829874ee7ac driver=raw_exec task_name=promtail error="rpc error: code = Unavailable desc = transport is closing"
2019-03-11T12:17:06.960+0800 [INFO ] client.gc: marking allocation for GC: alloc_id=6d3679d1-f40e-f2ce-080c-3829874ee7ac
2019/03/11 12:17:06.967491 [INFO] (runner) stopping
2019/03/11 12:17:06.967606 [INFO] (runner) received finish


here is log in the noamd server:

2019-03-11T12:17:06.785+0800 [ERROR] client.driver_mgr.raw_exec: error receiving stream from Stats executor RPC, closing stream: alloc_id=3228fc8d-dff6-7517-88f4-be81dbb47cdb driver=raw_exec task_name=promtail error="rpc error: code = Unavailable desc = transport is closing"
2019-03-11T12:17:06.791+0800 [INFO ] client.gc: marking allocation for GC: alloc_id=3228fc8d-dff6-7517-88f4-be81dbb47cdb
2019/03/11 12:17:06.798793 [INFO] (runner) stopping
2019/03/11 12:17:06.798926 [INFO] (runner) received finish

here is the state from nomad-ui:
![nomad1](https://user-images.githubusercontent.com/19262116/54100212-534cc000-43f8-11e9-9863-f23dbad73372.PNG)
![nomad2](https://user-images.githubusercontent.com/19262116/54100214-5647b080-43f8-11e9-8dbf-8a22bbead436.PNG)

here is the alloc dir:
![nomad3](https://user-images.githubusercontent.com/19262116/54100260-97d85b80-43f8-11e9-8856-46ebe2b7fdbd.PNG)

the job with allocid: 08011041 doesn.t exist in it.

when I click the start button:

logs in the nomad server:

2019/03/11 12:31:56.629972 [INFO] (runner) creating new runner (dry: false, once: false)
2019/03/11 12:31:56.630079 [INFO] (runner) creating watcher
2019/03/11 12:31:56.630197 [INFO] (runner) starting
2019/03/11 12:31:56.632365 [INFO] (runner) rendered "(dynamic)" =>"/var/lib/mos/nomad/alloc/6e879dbd-59c9-c9b1-ca76-a2923e9c396a/promtail/local/promtail.yml"
2019-03-11T12:31:56.633+0800 [INFO ] client.driver_mgr.raw_exec: starting task: driver=raw_exec

--- a/command/agent/consul/check_watcher.go
+++ b/command/agent/consul/check_watcher.go
@@ -126,7 +126,7 @@ func asyncRestart(ctx context.Context, logger log.Logger, task TaskRestarter, ev
 	if err := task.Restart(ctx, event, failure); err != nil {
 		// Restart errors are not actionable and only relevant when
 		// debugging allocation lifecycle management.
-		logger.Debug("error restarting task", "error", err,
+		logger.Debug("failed to restart task", "error", err,
 			"event_time", event.Time, "event_type", event.Type)
 	}
 }

--- a/command/agent/consul/check_watcher.go
+++ b/command/agent/consul/check_watcher.go
@@ -103,18 +103,32 @@ func (c *checkRestart) apply(ctx context.Context, now time.Time, status string) 
 		c.logger.Debug("restarting due to unhealthy check")
 
 		// Tell TaskRunner to restart due to failure
-		const failure = true
 		reason := fmt.Sprintf("healthcheck: check %q unhealthy", c.checkName)
 		event := structs.NewTaskEvent(structs.TaskRestartSignal).SetRestartReason(reason)
-		err := c.task.Restart(ctx, event, failure)
-		if err != nil {
-			// Error restarting
-			return false
-		}
+		go asyncRestart(ctx, c.logger, c.task, event)
 		return true
 	}
 
 	return false
+}
+
+// asyncRestart mimics the pre-0.9 TaskRunner.Restart behavior and is intended
+// to be called in a goroutine.
+func asyncRestart(ctx context.Context, logger log.Logger, task TaskRestarter, event *structs.TaskEvent) {
+	// Check watcher restarts are always failures
+	const failure = true
+
+	// Restarting is asynchronous so there's no reason to allow this
+	// goroutine to block indefinitely.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	if err := task.Restart(ctx, event, failure); err != nil {
+		// Restart errors are not actionable and only relevant when
+		// debugging allocation lifecycle management.
+		logger.Debug("error restarting task", "error", err,
+			"event_time", event.Time, "event_type", event.Type)
+	}
 }
 
 // checkWatchUpdates add or remove checks from the watcher

--- a/command/agent/consul/check_watcher_test.go
+++ b/command/agent/consul/check_watcher_test.go
@@ -3,12 +3,14 @@ package consul
 import (
 	"context"
 	"fmt"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/hashicorp/consul/api"
 	"github.com/hashicorp/nomad/helper/testlog"
 	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/testutil"
 	"github.com/stretchr/testify/require"
 )
 
@@ -34,6 +36,8 @@ type fakeCheckRestarter struct {
 	allocID   string
 	taskName  string
 	checkName string
+
+	mu sync.Mutex
 }
 
 // newFakeCheckRestart creates a new TaskRestarter. It needs all of the
@@ -54,6 +58,8 @@ func newFakeCheckRestarter(w *checkWatcher, allocID, taskName, checkName string,
 // Restarts are recorded in the []restarts field and re-Watch the check.
 //func (c *fakeCheckRestarter) Restart(source, reason string, failure bool) {
 func (c *fakeCheckRestarter) Restart(ctx context.Context, event *structs.TaskEvent, failure bool) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	restart := checkRestartRecord{
 		timestamp: time.Now(),
 		source:    event.Type,
@@ -69,11 +75,24 @@ func (c *fakeCheckRestarter) Restart(ctx context.Context, event *structs.TaskEve
 
 // String for debugging
 func (c *fakeCheckRestarter) String() string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
 	s := fmt.Sprintf("%s %s %s restarts:\n", c.allocID, c.taskName, c.checkName)
 	for _, r := range c.restarts {
 		s += fmt.Sprintf("%s - %s: %s (failure: %t)\n", r.timestamp, r.source, r.reason, r.failure)
 	}
 	return s
+}
+
+// GetRestarts for testing in a threadsafe way
+func (c *fakeCheckRestarter) GetRestarts() []checkRestartRecord {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	o := make([]checkRestartRecord, len(c.restarts))
+	copy(o, c.restarts)
+	return o
 }
 
 // checkResponse is a response returned by the fakeChecksAPI after the given
@@ -89,6 +108,8 @@ type fakeChecksAPI struct {
 	// responses is a map of check ids to their status at a particular
 	// time. checkResponses must be in chronological order.
 	responses map[string][]checkResponse
+
+	mu sync.Mutex
 }
 
 func newFakeChecksAPI() *fakeChecksAPI {
@@ -97,10 +118,14 @@ func newFakeChecksAPI() *fakeChecksAPI {
 
 // add a new check status to Consul at the given time.
 func (c *fakeChecksAPI) add(id, status string, at time.Time) {
+	c.mu.Lock()
 	c.responses[id] = append(c.responses[id], checkResponse{at, id, status})
+	c.mu.Unlock()
 }
 
 func (c *fakeChecksAPI) Checks() (map[string]*api.AgentCheck, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	now := time.Now()
 	result := make(map[string]*api.AgentCheck, len(c.responses))
 
@@ -349,4 +374,58 @@ func TestCheckWatcher_MultipleChecks(t *testing.T) {
 	if n := len(restarter3.restarts); n != 0 {
 		t.Errorf("expected check 3 to not be restarted but found %d:\n%s", n, restarter3)
 	}
+}
+
+// TestCheckWatcher_Deadlock asserts that check watcher will not deadlock when
+// attempting to restart a task even if its update queue is full.
+// https://github.com/hashicorp/nomad/issues/5395
+func TestCheckWatcher_Deadlock(t *testing.T) {
+	t.Parallel()
+
+	fakeAPI, cw := testWatcherSetup(t)
+
+	// If TR.Restart blocks, restarting len(checkUpdateCh)+1 checks causes
+	// a deadlock due to checkWatcher.Run being blocked in
+	// checkRestart.apply and unable to process updates from the chan!
+	n := cap(cw.checkUpdateCh) + 1
+	checks := make([]*structs.ServiceCheck, n)
+	restarters := make([]*fakeCheckRestarter, n)
+	for i := 0; i < n; i++ {
+		c := testCheck()
+		r := newFakeCheckRestarter(cw,
+			fmt.Sprintf("alloc%d", i),
+			fmt.Sprintf("task%d", i),
+			fmt.Sprintf("check%d", i),
+			c,
+		)
+		checks[i] = c
+		restarters[i] = r
+	}
+
+	// Run
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go cw.Run(ctx)
+
+	// Watch
+	for _, r := range restarters {
+		cw.Watch(r.allocID, r.taskName, r.checkName, r.check, r)
+	}
+
+	// Make them all fail
+	for _, r := range restarters {
+		fakeAPI.add(r.checkName, "critical", time.Time{})
+	}
+
+	// Ensure that restart was called exactly once on all checks
+	testutil.WaitForResult(func() (bool, error) {
+		for _, r := range restarters {
+			if n := len(r.GetRestarts()); n != 1 {
+				return false, fmt.Errorf("expected 1 restart but found %d", n)
+			}
+		}
+		return true, nil
+	}, func(err error) {
+		require.NoError(t, err)
+	})
 }

--- a/command/agent/consul/check_watcher_test.go
+++ b/command/agent/consul/check_watcher_test.go
@@ -9,6 +9,7 @@ import (
 	"github.com/hashicorp/consul/api"
 	"github.com/hashicorp/nomad/helper/testlog"
 	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/stretchr/testify/require"
 )
 
 // checkRestartRecord is used by a testFakeCtx to record when restarts occur
@@ -192,6 +193,28 @@ func TestCheckWatcher_Healthy(t *testing.T) {
 	if n := len(restarter2.restarts); n > 0 {
 		t.Errorf("expected check 2 to not be restarted but found %d:\n%s", n, restarter2)
 	}
+}
+
+// TestCheckWatcher_Unhealthy asserts unhealthy tasks are restarted exactly once.
+func TestCheckWatcher_Unhealthy(t *testing.T) {
+	t.Parallel()
+
+	fakeAPI, cw := testWatcherSetup(t)
+
+	check1 := testCheck()
+	restarter1 := newFakeCheckRestarter(cw, "testalloc1", "testtask1", "testcheck1", check1)
+	cw.Watch("testalloc1", "testtask1", "testcheck1", check1, restarter1)
+
+	// Check has always been failing
+	fakeAPI.add("testcheck1", "critical", time.Time{})
+
+	// Run
+	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
+	defer cancel()
+	cw.Run(ctx)
+
+	// Ensure restart was called exactly once
+	require.Len(t, restarter1.restarts, 1)
 }
 
 // TestCheckWatcher_HealthyWarning asserts checks in warning with

